### ISSUE TITLE
Training Draft for Large models (e.g 70B)

### ISSUE THREAD

**Issue by Ofir408** [NVIDIA/Model-Optimizer#593]
Hi,
I'm training an eagle3 draft model for speculative decoding for a large model such as Llama-70B:
`
bash train_eagle3_and_export.sh \
  --base_model /opt/ml/model/base_models/Llama3.3-70B-instruct \
  --data /opt/ml/model/data/train.jsonl \
  --num_gpu 8
`
 From what I understand, this requires using the offline training mode. However, when I tried training without saving hidden states, I ran into OOM issues. I**t looks like ~97% of the Llama-70B parameters remain trainable even after [converting to an Eagle draft model** ([https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/1aaa77d1af66ebedf072fd4327c461a192d9dec9/examples/speculative_decoding/main.py#L212](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/1aaa77d1af66ebedf072fd4327c461a192d9dec9/examples/speculative_decoding/main.py#L212)).

How is this expected? does the final conversion into an Eagle model happens after training, using the export_hf_checkpoint.py script. Does this mean the model remains mostly unfrozen during training?

Additionally, I attempted dumping hidden states to disk using the run_hf_compute_hiddens_dp script you provided, but the process is extremely slow — around 4 days for only 120K examples on 8×A100 80GB.

Is there a more efficient workflow or recommended approach for training a draft model for Llama-70B?

Thanks!

**Comment by Ofir408** [comment]
same for gpt-oss-20b. can you please help? @ChenhanYu and @yeyu-nvidia ? 
Thanks!

**Comment by yeyu-nvidia** [comment]
Hi @Ofir408, can you check model.eagle_freeze_base_model after mtsp.convert? It should be set to True. If so, the base model is frozen here https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/modelopt/torch/speculative/plugins/transformers.py#L668

**Comment by h-guo18** [comment]
Hi @Ofir408 , 

Although some base model's param have `requres_grad=True`, it should not take additional memory since the base model is runned under `torch.no_grad()`context. ([code link](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/modelopt/torch/speculative/plugins/transformers.py#L674))

To train with llama-70B base model, I recommend enable FSDP by addingthe following argument in `launch_train.sh`:

```
CMD="accelerate launch $MULTI_GPU --mixed_precision bf16 main.py \
    --mode $MODE \
    --model_name_or_path $MODEL \
    --training_seq_len $TRAINING_SEQ_LEN \
    --dataloader_drop_last True \
    --bf16 True \
    --output_dir $OUTPUT_DIR \
    --num_train_epochs $NUM_EPOCHS \
    --per_device_train_batch_size $TRAIN_BS \
    --per_device_eval_batch_size $TRAIN_BS \
    --gradient_accumulation_steps 1 \
    --do_eval False \
    --eval_accumulation_steps 1 \
    --save_strategy steps \
    --save_steps $SAVE_STEPS \
    --learning_rate $LR \
    --weight_decay 0.0 \
    --warmup_steps 100 \
    --lr_scheduler_type linear \
    --logging_steps 100 \
    --tf32 True \
    --data_path $DATA \
    --disable_tqdm $DISABLE_TQDM \
    --ar_validate_steps $AR_VALIDATE_STEPS \
    $VLM_ARGS \
    $OFFLINE_TRAINING_ARGS \
    $SPECULATIVE_ARGS \
    --fsdp 'full_shard auto_wrap'            <-----  #NOTICE: newly added to enable fsdp
"
```

Then 1k training seq len + online training should work on 8x80GB machine.

As for data dumping, the speed is related to disk bandwidth. I recommend trying faster disk if available, or use online training instead.
